server/app/services/program/ProgramServiceImpl.java

```diff
@@ -708,6 +710,50 @@ public ProgramDefinition setProgramQuestionDefinitionOptionality(
     }
   }
 
+  @Override
+  @Transactional
+  public ProgramDefinition setProgramQuestionDefinitionPosition(
+      long programId, long blockDefinitionId, long questionDefinitionId, int newPosition)
+      throws ProgramNotFoundException, ProgramBlockDefinitionNotFoundException,
+          ProgramQuestionDefinitionNotFoundException, InvalidQuestionPositionException {
+    ProgramDefinition programDefinition = getProgramDefinition(programId);
+    BlockDefinition blockDefinition = programDefinition.getBlockDefinition(blockDefinitionId);
+
+    ArrayList<ProgramQuestionDefinition> questions =
+        new ArrayList<>(blockDefinition.programQuestionDefinitions());
+
+    if (newPosition < 0 || newPosition >= questions.size()) {
+      throw new InvalidQuestionPositionException(newPosition, questions.size());
+    }
+
+    // move question to the new position
+    ProgramQuestionDefinition question = null;
+    for (Iterator<ProgramQuestionDefinition> it = questions.iterator(); it.hasNext(); ) {
+      question = it.next();
+      if (question.id() == questionDefinitionId) {
+        it.remove();
+        break;
+      }
+    }
+    if (question == null) {
+      throw new ProgramQuestionDefinitionNotFoundException(
+          programId, blockDefinitionId, questionDefinitionId);
+    }
+    questions.add(newPosition, question);
+
+    try {
+      return updateProgramDefinitionWithBlockDefinition(
+          programDefinition,
+          blockDefinition.toBuilder()
+              .setProgramQuestionDefinitions(ImmutableList.copyOf(questions))
+              .build());
+    } catch (IllegalPredicateOrderingException e) {
+      // Changing a question position within block should not affect predicates.
```

Comment:
nit: maybe add "Changing a question position within block should not affect predicates because predicates cannot depend on questions within the same block."

Reply:
Done.